core/shared/src/main/scala/cats/effect/cps/AsyncAwait.scala

```diff
@@ -88,49 +85,81 @@ object AsyncAwaitDsl {
+        // AsyncAwaitStateMachine (as it indicates the macro-expansion of a nested async region)
+        def rec(t: Tree): Iterator[c.Tree] = Iterator(t) ++ t.children.filter(_ match {
+          case ClassDef(_, _, _, Template(List(parent), _, _)) if parent.symbol == stateMachineSymbol =>
+            false
+          case _ =>
+            true
+        }).flatMap(rec(_))
+
+        // Checking each local `await` call to ensure that it matches the `F` in `async[F]`
+        val effect = F.actualType.typeArgs.head
+
+        rec(body.tree).foreach {
+          case tt @ c.universe.Apply(TypeApply(_, List(awaitEffect, _)), fun :: Nil) if tt.symbol == awaitSym =>
+            // awaitEffect is the F in `await[F, A](fa)`
+            if (!(awaitEffect.tpe <:< effect)){
+              c.abort(fun.pos, s"Expected await to be called on ${effect}, but was called on ${fun.tpe.dealias}")
+            }
+          case _ => ()
+        }
+
         def mark(t: DefDef): Tree = {
           c.internal
             .asInstanceOf[{
               def markForAsyncTransform(
                   owner: Symbol,
                   method: DefDef,
                   awaitSymbol: Symbol,
-                  config: Map[String, AnyRef]
-              ): DefDef
+                  config: Map[String, AnyRef])
+                  : DefDef
             }]
             .markForAsyncTransform(
               c.internal.enclosingOwner,
               t,
               awaitSym,
-              Map.empty
-            )
+              Map.empty)
         }
+
         val name = TypeName("stateMachine$async")
+
         // format: off
         val tree = q"""
-          final class $name(dispatcher: _root_.cats.effect.std.Dispatcher[${c.prefix}._AsyncContext], callback: _root_.cats.effect.cps.AsyncAwaitDsl.AwaitCallback[${c.prefix}._AsyncContext]) extends _root_.cats.effect.cps.AsyncAwaitStateMachine(dispatcher, callback) {
-            ${mark(q"""override def apply(tr$$async: _root_.cats.effect.cps.AsyncAwaitDsl.AwaitOutcome[${c.prefix}._AsyncContext]): _root_.scala.Unit = ${body}""")}
+          final class $name(dispatcher: _root_.cats.effect.std.Dispatcher[$effect], callback: _root_.cats.effect.cps.AsyncAwaitDsl.AwaitCallback[$effect]) extends _root_.cats.effect.cps.AsyncAwaitStateMachine(dispatcher, callback)(${F}) {
+            ${mark(q"""override def apply(tr$$async: _root_.cats.effect.cps.AsyncAwaitDsl.AwaitOutcome[$effect]): _root_.scala.Unit = $body""")}
           }
-          ${c.prefix}._AsyncInstance.flatten {
-            _root_.cats.effect.std.Dispatcher[${c.prefix}._AsyncContext].use { dispatcher =>
-              ${c.prefix}._AsyncInstance.async_[${c.prefix}._AsyncContext[AnyRef]](cb => new $name(dispatcher, cb).start())
+          ${F}.flatten {
+            _root_.cats.effect.std.Dispatcher[$effect].use { dispatcher =>
+              ${F}.async_[$name#FF[AnyRef]](cb => new $name(dispatcher, cb).start())
```

Comment:
This is somewhat infuriating, but couldn't figure for the life of me how to apply an argument to the `effect` type value (which represents the polymorphic F). So instead, added a type member to `AsyncAwaitStateMachine` so that I'm able to reference it using a type projection ... happy for it to be changed if you know how to.

Reply:
LOL wow that's annoying. I'm not that worried about it though. We can also iterate on it later.